designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,401 @@
+    edit(context) {
+        const prettier = require("prettier");
+        return prettier.format(context.sourceText, { semi: false, parser: "babylon" });
+    }
+}
+```
+
+An example using [Recast](https://github.com/benjamn/recast) (example is from Recast README):
+
+```js
+{
+    type: "text",
+    edit(context) {
+        const recast = require("recast");
+        const ast = recast.parse(context.sourceText);
+
+        // Grab a reference to the function declaration we just parsed.
+        var add = ast.program.body[0];
+
+        // Make sure it's a FunctionDeclaration (optional).
+        var n = recast.types.namedTypes;
+        n.FunctionDeclaration.assert(add);
+
+        // If you choose to use recast.builders to construct new AST nodes, all builder
+        // arguments will be dynamically type-checked against the Mozilla Parser API.
+        var b = recast.types.builders;
+
+        // This kind of manipulation should seem familiar if you've used Esprima or the
+        // Mozilla Parser API before.
+        ast.program.body[0] = b.variableDeclaration("var", [
+            b.variableDeclarator(add.id, b.functionExpression(
+                null, // Anonymize the function expression.
+                add.params,
+                add.body
+            ))
+        ]);
+
+        // Just for fun, because addition is commutative:
+        add.params.push(add.params.shift());
+
+        // return the modified code
+        return recast.print(ast).code;
+    }
+}
+```
+
+Any JavaScript utility that can work on a string and return a string can be used inside a `text` style edit.
+
+Because the style edits have a defined order inside of a style editor, end-users can prescribe any number of steps to edit styles in source code. For example, you might want to run a Recast edit and then a Prettier edit:
+
+```js
+// style editor
+module.exports = {
+    meta: {
+        name: "ES5 Style",
+        description: "A style editor just for ES5"
+    },
+    edits: [
+        require("./edits/recast"),
+        require("./edits/prettier")
+    ]
+}
+```
+
+This ability to order individual edits means that end-users could potentially run Prettier to get a baseline format for their source code and then run additional fixes to tweak the style in whichever way they like. Or they could perform a bunch of AST transforms using Recast and then apply Prettier to format everything.
+
+### Style Editor Composition
+
+Because style editors are just objects, end-users can compose new style editors from existing ones. For example, if someone wanted to use an Airbnb style editor but tweak the indentation scheme a bit, they could create a new style editor like this:
+
+```js
+const airbnbStyle = require("eslint-plugin-airbnb").styles.default;
+
+module.exports = {
+    meta: {
+        name: "Modified Airbnb Style",
+        description: "A style editor based on Airbnb"
+    },
+    edits: [
+        ...airbnbStyle.edits,
+        require("./edits/indentation")
```

Comment:
It seems like style editors would often conflict with each other, similar to how autofixing one rule occasionally creates linting errors for another rule. Autofixes from rules are applied multiple times (through multipass autofixing), which addresses the vast majority of conflicts. Would style editors also be applied multiple times?

Example of a conflict:

```js
module.exports = {
    edits: [
        require("./edits/indentation"),
        require("./edits/add-linebreaks-if-lines-are-too-long")
    ]
};
```

The `add-linebreaks-if-lines-are-too-long` style editor wouldn't know the user's indentation preferences, and presumably it wouldn't want to bundle indentation logic, so the user's code would end up with some indentation errors after linebreaks are added. It's likely that the user wouldn't notice this issue because no linting errors would be reported at the end (assuming the user disables the `indent` rule in favor of just using style editors).

Note that switching the order of the style editors wouldn't help here. If user decided to run the indentation editor after the linebreak-adding editor, then adding indentation to a line could cause it to become "too long" such that it would be appropriate to have linebreaks added, but the linebreak-adding editor would have already run.

More generally, I have some doubts about how composable style editors would be in practice. It's true that the sequence of two valid style editors is also a valid style editor, but it seems like it would be difficult to ensure that the intended aspects of the original style editors are still preserved after combining them.

Reply:
You are correct in that this is possible. The difference from the current situation is that fixing the conflict is solely on the end-user and doesn't involve us (compared to now, where we end up trying to figure out how to get two rules to not conflict).

In the specific case you mention, the options for resolving the problem are:

1. Choose just one of the edits.
2. Choose a different order of the edits based on your priority.
3. Remove both edits and write a new one that does what you want.

The most common scenario in my mind is something like this: let's say there is an `indent` edit that I've provided in a package. By default it indents a `switch` block like this:

```js
switch (foo) {
case 1: return 1;
case 2: return 2;
case 3: return 3;
}
```

But what you want is this:

```js
switch (foo) {
    case 1: return 1;
    case 2: return 2;
    case 3: return 3;
}
```

Then you would apply the `indent` edit first and then apply an edit that just targeted indenting those `case` statements one level deeper.

And we can provide `styleOptions` that are passed into the style editors. You could very well have indentation info in that if necessary.

Just to reiterate: the upside here is that figuring out how to get the right code format output is on the end-user, and they can use any tool to do that.